lib/src/library_analyzer.dart

```diff
@@ -0,0 +1,185 @@
+    if (packageMap == null) {
+      throw new StateError('An error occurred getting the package map.');
+    }
+
+    return new PackageMapUriResolver(
+        PhysicalResourceProvider.INSTANCE, packageMap);
+  }
+
+  Future<List<String>> _scanUri(String libUri) async {
+    Uri uri = Uri.parse(libUri);
+    String package = uri.path.split('/').first;
+    String fullPath = _packageResolver.resolveAbsolute(uri).fullName;
+    String relativePath =
+        p.join('lib', libUri.substring(libUri.indexOf('/') + 1));
+    if (fullPath.endsWith('/$relativePath')) {
+      String packageDir =
+          fullPath.substring(0, fullPath.length - relativePath.length - 1);
+      List<String> libs = _parseLibs(package, packageDir, relativePath);
+      _cachedLibs[libUri] = libs;
+      return libs;
+    } else {
+      return [];
+    }
+  }
+
+  Future<Map<String, List<String>>> _scanPackage(
+      String package, String packageDir) async {
+    await _init();
+    Map<String, List<String>> results = {};
+    List<String> dartFiles = await listFiles(packageDir, endsWith: '.dart');
+    List<String> mainFiles = dartFiles
+        .where((path) => path.startsWith('lib/') || path.startsWith('bin/'))
+        .toList();
+    for (String relativePath in mainFiles) {
+      String uri = _toUri(package, relativePath);
+      if (!_cachedLibs.containsKey(uri)) {
+        _cachedLibs[uri] = _parseLibs(package, packageDir, relativePath);
+      }
+      results[uri] = _cachedLibs[uri];
+    }
+    return results;
+  }
+
+  List<String> _parseLibs(
+      String package, String packageDir, String relativePath) {
+    String fullPath = p.join(packageDir, relativePath);
+    LibraryElement lib = _getLibraryElement(fullPath);
+    if (lib == null) return [];
+    Set<String> refs = new Set();
+    lib.importedLibraries.forEach((le) {
+      refs.add(_normalizeLibRef(le.librarySource.uri, package, packageDir));
+    });
+    lib.exportedLibraries.forEach((le) {
+      refs.add(_normalizeLibRef(le.librarySource.uri, package, packageDir));
+    });
+    refs.remove('dart:core');
+    return refs.toList()..sort();
+  }
+
+  LibraryElement _getLibraryElement(String path) {
+    Source source = new FileBasedSource(new JavaFile(path));
+    if (_context.computeKindOf(source) == SourceKind.LIBRARY) {
+      return _context.computeLibraryElement(source);
+    }
+    return null;
+  }
+}
+
+String _normalizeLibRef(Uri uri, String package, String packageDir) {
+  if (uri.isScheme('file')) {
+    String relativePath = p.relative(p.fromUri(uri), from: packageDir);
+    return _toUri(package, relativePath);
+  } else if (uri.isScheme('package') || uri.isScheme('dart')) {
+    return uri.toString();
+  }
+
+  throw "not supported - $uri";
+}
+
+String _toUri(String package, String relativePath) {
+  if (relativePath.startsWith('lib/')) {
+    return 'package:$package/${relativePath.substring(4)}';
+  } else {
+    return 'path:$package/$relativePath';
```

Comment:
I have one argument for the current way:
https://github.com/kevmoo/pana.dart/pull/4/files#diff-5037fb91fae4e421431eae8fd216c9cbR161

With the current impl, both the keys and the library values follow the same pattern, regardless of their context (e.g. whether this is the source package or a dependent one). This allows allows simple caching, and if we remove it, the caching logic needs to become more complex (e.g. need to consider/transform the values in a different context).